coding/src/zoda.rs

```diff
@@ -0,0 +1,660 @@
+//! This module implements the [ZODA](https://eprint.iacr.org/2025/034) coding scheme.
+//!
+//! At a high level, the scheme works like any other coding scheme: you start with
+//! a piece of data, and split it into shards, and a commitment. Each shard can
+//! be checked to belong to the commitment, and, given enough shards, the data can
+//! be reconstructed.
+//!
+//! What makes ZODA interesting is that upon receiving and checking one shard,
+//! you become convinced that there exists an original piece of data that will
+//! be reconstructable given enough shards. This fails in the case of, e.g.,
+//! plain Reed-Solomon coding. For example, if you give people random shards,
+//! instead of actually encoding data, then when they attempt to reconstruct the
+//! data, they can come to different results depending on which shards they use.
+//!
+//! Ultimately, this stems from the fact that you can't know if your shard comes
+//! from a valid encoding of the data until you have enough shards to reconstruct
+//! the data. With ZODA, you know that the shard comes from a valid encoding as
+//! soon as you've checked it.
+//!
+//! # Variant
+//!
+//! ZODA supports different configurations based on the coding scheme you use
+//! for sharding data, and for checking it.
+//!
+//! We use the Reed-Solomon x Hadamard variant of ZODA: in essence, this means
+//! that the shards are Reed-Solomon encoded, and we include additional checksum
+//! data which does not help reconstruct the data.
+//!
+//! ## Deviations
+//!
+//! In the paper, a sample consists of rows chosen at random from the encoding of
+//! the data. With multiple participants receiving samples, they might receive
+//! overlapping samples, which we don't want. Instead, we shuffle the rows of
+//! the encoded data, and each participant receives a different segment.
+//! From that participant's perspective, they've received a completely random
+//! choice of rows. The other participants' rows are less random, since they're
+//! guaranteed to not overlap. However, no guarantee on the randomness of the other
+//! rows is required: each sample is large enough to guarantee that the data
```

Comment:
I suspect this will be the primary claim people poke at. I wonder if there is some way we can add some intuition around a malicious dealer not being able to circumvent this?